### PR TITLE
Feature/st 169 alternative cardinal flow

### PR DESCRIPTION
**Pull request by james-simpson-ST**

### CODE REVIEW

**Comment by pgs-gmartowski** [comment]
Couple of comments, but generally lgtm.

File: example/immediate.html [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+<body>
+  <form id="st-form" class="example-form" action="receipt.html">
+    <div id="st-notification-frame" class="example-form__group"></div>
+    </div>
+    <div id="st-control-frame" class="example-form__group"></div>
+  </form>
+  <script type="text/javascript" src="st.js"></script>
+  <script>
+    (function () {
+      const parsedUrl = new URL(window.location.href);
```
[the remark this answers is not in the document]
Comment (james-simpson-ST) [reply example/immediate.html]: Done - don't need parsed values in immediate so removing

File: src/ST.ts [hunk 2]
```diff
@@ -19,6 +19,9 @@ class ST {
   private tokenise: boolean;
   private styles: IStyles;
   private componentIds: any;
+  private submitOnSuccess: boolean;
+  private submitOnError: boolean;
+  private submitFields: string[];
```
[the remark this answers is not in the document]
Comment (james-simpson-ST) [reply src/ST.ts]: Done

File: src/core/classes/CommonFrames.class.ts [hunk 3]
```diff
@@ -61,10 +83,16 @@ export default class CommonFrames extends RegisterFrames {
...
     targets.map((item, index) => {
       const itemToChange = document.getElementById(item);
-      itemToChange.appendChild(fields[index]);
+      if (fields[index]) {
+        itemToChange.appendChild(fields[index]);
+      }
     });
   }
 
+  private shouldLoadNotificationFrame() {
```
[the remark this answers is not in the document]
Comment (james-simpson-ST) [reply src/core/classes/CommonFrames.class.ts]: When I changed this it no longer detected shouldLoadNotificationFrame as a function on "this"? Not sure why

File: src/core/classes/CommonFrames.class.ts [hunk 4]
```diff
@@ -79,4 +107,23 @@ export default class CommonFrames extends RegisterFrames {
...
   }
+
+  private onTransactionComplete(data: any) {
+    if (
+      (this.submitOnSuccess &&
+        data.errorcode === '0' &&
+        ['AUTH', 'CACHETOKENISE'].includes(data.requesttypedescription)) ||
+      (this.submitOnError && data.errorcode !== '0')
+    ) {
+      const form: HTMLFormElement = document.getElementById(Selectors.MERCHANT_FORM_SELECTOR) as HTMLFormElement;
```
[the remark this answers is not in the document]
Comment (james-simpson-ST) [reply src/core/classes/CommonFrames.class.ts]: Done - moved to static class variable

File: src/core/classes/StTransport.class.ts [hunk 5]
```diff
@@ -56,7 +56,11 @@ export default class StTransport {
     return this.fetchRetry(this.gatewayUrl, {
       ...StTransport.DEFAULT_FETCH_OPTIONS,
       body: this._codec.encode(requestObject)
-    }).then(this._codec.decode);
+    })
+      .then(this._codec.decode)
+      .catch(e => {
```
[the remark this answers is not in the document]
Comment (james-simpson-ST) [reply src/core/classes/StTransport.class.ts]: Done

File: src/core/integrations/CardinalCommerce.ts [hunk 6]
```diff
@@ -110,6 +126,17 @@ export class CardinalCommerce {
         threedresponse: jwt
       });
     } else {
+      const responseData = {
```
[the remark this answers is not in the document]
Comment (james-simpson-ST) [reply src/core/integrations/CardinalCommerce.ts]: Done and updated here and StCodec to use it :)

File: src/core/classes/CommonFrames.class.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (pgs-gmartowski) [reply src/core/classes/CommonFrames.class.ts]: That's weird :/ Ok, then we can leave it as:
```
private shouldLoadNotificationFrame() {
  return !(this.submitOnError && this.submitOnSuccess);
  }
```

**Pull request merged**